fs: add Resumer interface

mcalman opened the pull request:
Added an interface for resuming failed uploads. Later on this interface
can be implemented by any backend where resumes are possible.

#87

<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Adds an optional interface for resuming uploads that can later be implemented by any backend where resuming uploads is possible. Once implemented by a backend, this will allow uploads to continue after a failure, even when restarting Rclone operations.

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/87
https://forum.rclone.org/t/drive-resume-resumable-upload-implementation/18289/4
https://forum.rclone.org/t/chunker-resume-upload-contribution/18571

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

mcalman committed:
chunker: implement resumer interface

Fixes #5154

mcalman committed:
Add resume feature

Added an interface and machinery for resuming failed uploads.
Implemented this interface in the local backend.
Later on it can be implemented by any supporting backend.

Fixes #87

ncw commented:
This looks very nice :-)

I'm just in the middle of making the 1.53 release - I will review in detail after that is done.

ncw commented on fs/operations/operations.go:
```diff
@@ -343,6 +344,112 @@ func CommonHash(fa, fb fs.Info) (hash.Type, *fs.HashesOption) {
+		//
+		// SetID takes the passed resume ID and Fingerprint of the object and stores it in
+		// --cache-dir so that future Copy operations can resume the upload if it fails
+		func(ID string) (err error) {
```
This should be a separate function too I think, so the closure you pass as SetID: is only a line or two long.

ncw commented on fs/operations/operations.go:
```diff
@@ -343,6 +344,112 @@ func CommonHash(fa, fb fs.Info) (hash.Type, *fs.HashesOption) {
+		// --cache-dir so that future Copy operations can resume the upload if it fails
+		func(ID string) (err error) {
+			rootCacheDir := filepath.Join(config.CacheDir, "resume")
+			if filepath.Ext(rootCacheDir) != "" {
```
I don't understand why you are looking at `filepath.Ext` here? Surely the last path item is "resume" isn't it?

ncw commented on fs/operations/operations.go:
```diff
@@ -343,6 +344,112 @@ func CommonHash(fa, fb fs.Info) (hash.Type, *fs.HashesOption) {
+			defer func() {
+				_ = cacheFile.Close()
+			}()
+			// Check how many resume cache files already exist and remove oldest if more than limit
```
Do we want to do this on *every* file we upload?

ivandeex replied:
compare e.g. with _curl_ tool, which activates resume logic only when user adds _--continue-at_ on command line (https://curl.se/docs/manpage.html, google "curl continue")

mcalman replied:
I made a slight tweak so that the cache is only cleaned once per file uploaded. Previously, if a backend had to repeatedly update the cache for a single file (e.g. Chunker updating the cache with each successfully uploaded chunk) the cache was being unnecessarily cleaned during each update.

@ncw Do you have any suggestions for how we might further mitigate how often the cache is cleaned?

@ivandeex This is certainly something I can look into adding. Would we want such an option to only control whether or not to attempt a resume on an interrupted upload, or would we also want to only conditionally write the necessary information to the cache to make a later resume possible? Either with an option or maybe only for files above a certain size where resumes are more useful?

ivandeex replied:
`--resume-larger` satisfies the request.

> Would we want such an option to only control whether or not to attempt a resume on an interrupted upload, or would we also want to only conditionally write the necessary information to the cache to make a later resume possible?

I think this bit will become clear when you proceed to adding unit tests (resume normally or resume after corrupted source/destination) and making them pass.

ncw replied:
Once per file seems OK thinking some more about it since we are only going to activate this for large files probably which take lots of time. I was thinking we'd be doing it for every file which might be 100/s but I think that is incorrect.

ncw commented on fs/operations/operations.go:
```diff
@@ -343,6 +344,112 @@ func CommonHash(fa, fb fs.Info) (hash.Type, *fs.HashesOption) {
+			return nil
+		}}
+
+		// Check to see if a resume cache file exists for the file currrently being uploaded
```
Another candidate for a function?

ncw commented on fs/operations/operations.go:
```diff
@@ -428,6 +535,10 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
 						}
 						options := []fs.OpenOption{hashOption}
+						// Prepends OptionResume if it was set
+						if resumeOpt != (&fs.OptionResume{}) {
```
This could be `resumeOpt != nil` if you used the pointer above.

ncw commented on fs/operations/operations.go:
```diff
@@ -428,6 +535,10 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
 							wrappedSrc = NewOverrideRemote(src, remote)
 						}
 						options := []fs.OpenOption{hashOption}
+						// Prepends OptionResume if it was set
```
I don't think you need to prepend it, append will do like the other options.

ncw commented on fs/operations/operations.go:
```diff
@@ -492,6 +609,33 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
+}
+
+func unmarshalResumeJSON(ctx context.Context, data []byte) (fprint, id string, err error) {
+	if data == nil || len(data) < 2 || data[0] != '{' || data[len(data)-1] != '}' {
```
You don't need to do these tests `json.Unmarshal` will do them.

ncw reviewed (commented):
v1.53 is out the door! Apologies for the delay in reviewing.

I think this is a good start - thank you :-)

I wrote lots of comments inline.

Can you factor the ~100 extra lines in `operations.Copy` so we try to keep things readable.

We'll need to write tests too, so having functions we can test individually will be very useful :-)

ivandeex commented:
Related to https://github.com/rclone/rclone/issues/4798 (multi-thread upload interface)

ivandeex commented:
@mcalman 
Your PR has been rebased on current rclone master.
Please  **pull** results from github. **Don't** force-push now.
Please address remaining review comments too.
Thanks.

ivandeex commented:
@mcalman 
Please address inline review comments and flag us in a comment when you are ready.

mcalman commented:
@ivandeex @ncw All inline comments have been addressed except for this one which I'd like to hear your feedback on. https://github.com/rclone/rclone/pull/4547#discussion_r570464650

ivandeex commented on backend/chunker/chunker.go:
```diff
@@ -1352,6 +1352,20 @@ func (f *Fs) Purge(ctx context.Context, dir string) error {
+// the point the file should be resumed from or an error.
+//
+// This command will chain to `resume` from wrapped remote.
+// Can later be replaced by an implementaiton of `resume` for chunker
```
@mcalman
we can't just chain to the lower backend in a general case. if the file is chunked it's remote chains to a small metadata (and to nothing if metadata is disabled). if it's non-chunked, it can become chunked after resume, but we can't predict it given arguments here.

i suggest that you either return unsupported error or remove interface from chunker.

@ncw 
you requested to add resume support to all wrapping backends. i'm not sure how this can be done for chunker.

ivandeex replied:
let's add the opt-out clause for chunker

ivandeex commented:
@ncw
the flag can be named `--resume-larger <SizeSuffix>` and it will attempt resuming only if already uploaded size exceeds the parameter
or it can mimic curl and be just `--continue`, meaning "resume anything with non-zero already upliaded size"
I'm in favor of parameterized flag.
What do you think??

ivandeex commented:
This pull d4sperately lacks tests.
Tests will need at least one backend suppirting resume
I can only think of the local fs in that role (using file Open in append mode)

ivandeex commented:
If (and only if!) you want to test resume with chunker, you'll have to implement resume in chunker.

Consequently you'll want to add a new per-transaction control chunk to save info about partial hash(es) and probably hashes of uploaded chunks. Then you'll want to add code that selects transaction to resume given partial hash and upliaded size so far. Maybe select the "best" partial transaction (when _rename_ is fast) or just pick a single partial txn ID (when it's slow).

Bottomline: this will create metadata v3 and combinatorically grow number of tested combinations. So if the answer is "yes I want", I propose to commit this and the other chunker pull on a branch (after succesful reviews). This will create a beta release for beta-testing. Later we can  merge both pulls from the branch on master together using a single metadata version number.

ivandeex commented:
specific notes about cache:
- who decides to copy source to upload cache, backend or generic code in operations.go?
- if a source is "stable", can we avoid copying to cache and keep only json?
- if a source is "dynamic", can we force copying to cache?

ivandeex commented:
a note why i care about resumable checksums (if supported by a backend).
consider this scenario:
- user uploads a large file
- network broken, upliad canceled
- source file is changed or another attempt changes partial upliad on target
- user asks to resume a file
- rclone resumes (here we could have checked validity of partial upload)
- after some hours rclone finds that fingerprint is wrong
- abort, retry

ivandeex commented:
Another note, which relates simulatenously to the upload cache size, rolling hash and specific implementation of Resumer in the chunker backend (it can be true for the Local backend true, not sure about others).

In case of chunker particularly, upload cache usage can be somewhat decreased because already uploaded chunks are isolated remotely and marked by a "txn id", we just need a way to re-check them, and this is where the rolling hash can help us.

Again, implementation of Resumer for chunker is absolutely OUT of scope for this pull request,
but providing a framework, mechanisms, unit tests - IN.

ivandeex commented:
@mcalman I will re-request review from from ncw after we clear current round.

ivandeex commented:
Related to https://forum.rclone.org/t/http-upload-contribution/13271/38
Related to #3842 and #3850

ivandeex commented on backend/chunker/chunker.go:
```diff
@@ -1465,6 +1465,20 @@ func (f *Fs) Purge(ctx context.Context, dir string) error {
+	if do == nil {
+		return 0, fs.ErrorCantResume
+	}
+	return do(ctx, remote, ID)
```
@mcalman
Please also add `UnimplementableFsMethods: []string{..., "Resume"},` to the `chunker_test.go`, and we can mark this bit as resolved

ivandeex replied:
I think you can completely remove the `func Resume()` stub from `chunker.go`. Let's address it in another PR.
Testing resume interface with chunker is not a low hanging fruit, though it looks tempting.
I think `backend/local` is a better candidate for that.

ivandeex replied:
@mcalman
I think you can completely remove the `func Resume()` stub from `chunker.go`.
Testing the resume interface with chunker is not as simple as it may seem, though tempting. Let's address it in another PR.
I think `backend/local` is a better candidate for that.
Please also add `UnimplementableFsMethods: []string{..., "Resume"},` to the `chunker_test.go` so we can mark this thread as resolved.

ivandeex replied:
@mcalman
I think you can completely remove the `func Resume()` stub from `chunker.go`.
Testing the resume interface with chunker is not as simple as it may seem, though tempting. Let's address it elsewhere.
I think `backend/local` is the best candidate for unit tests here.
Please also add `UnimplementableFsMethods: []string{..., "Resume"},` to the `chunker_test.go` so we can mark this thread as resolved.

ivandeex commented:
@mcalman
Please squash commits and rebase on current master.

ivandeex commented:
@mcalman 
Please add tests, either in the form of synthetic tests against interface functions, or optionally using backend/local as the simplest possible implementation of Resumer.

ivandeex commented:
@ncw 
This PR will need some architecture coordination with TUS uploader #3850

ivandeex commented:
@ncw 
Please have a look at this.

Note that it additionally implements interface contracts stated in #5154 and in the discussion above (I hid all implemented comments, only the relevant ones are visible by default).
Also note that PR #5185 (implement Resumer in Chunker) depends on this one.

This PR will need some architecture coordination with #3850 (TUS uploader) and competing PR #5299 (ResumableUpload), which can be done only by you.

ivandeex commented:
@ncw 
Please have a look at this.

Note that in addition to your inline comments it implements interface contracts stated in #5154 and in the discussion above (I hid all outdated comments, only the relevant ones are visible by default).
Also note that PR #5185 (implement Resumer in Chunker) depends on this one.

This PR will need some architecture coordination with #3850 (TUS uploader) and competing PR #5299 (ResumableUpload), which can be done only by you.

innovate-invent commented on fs/fs.go:
```diff
@@ -932,6 +940,13 @@ func (ft *Features) WrapsFs(f Fs, w Fs) *Features {
+type Resumer interface {
+	// Resume checks whether the (remote, ID) pair is valid and returns
+	// the point the file should be resumed from or an error.
+	Resume(ctx context.Context, remote, ID, hashName, hashState string) (Pos int64, err error)
```
I don't think I agree with the idea that a hash should be generally provided by backends. That means that all backends must be able to provide a hash.

ivandeex replied:
No. The proper supports that chunker feature but other backends are not required to support it.

from https://github.com/rclone/rclone/pull/4547#issuecomment-786889739
> if backend does not support it, its interface will return a nil accrued hashsum

innovate-invent replied:
but the hash is in the interface signature

ivandeex replied:
It's in signature but it can be nil.

ivandeex replied:
> resumable uploads via a different mechanism.

This interface is generic enough but optional. Backends are free to ignore it. Please take a closer look at the source code.

ivandeex replied:
Here is quick synopsis of the code.
There are three parts here: the backend, the resume proper which keeps local cache of partial uploads and generic interface between them.
The flow is:
- During upload the resume proper takes snapshot by requesting supported info from backend.
- If later user decides to attempt resuming, the proper performs cache lookup and queries backend providing available info and a callback.
- The backend picks up supported pieces from the provided info, matches it with upstream, chooses resume identifier and calks back to the proper
- The resume aborts or continues

Notes: Rolling hashsums are just part of matched info along with size or name. They are generally optional. They can be supported by local or chunker backends. They provide additional data safety against broken uploads. There must be a good reason to drop them.

innovate-invent replied:
> resume proper which keeps local cache of partial uploads

Do you mean the cache is not stored on the remote? You can't assume that there is a local filesystem available to rclone.

> resume proper takes snapshot by requesting supported info from backend

rclone should make effort to support resumable uploads regardless of if the backend supports it or not. This can be done with a configurable fallback backend for intermediate files

> There must be a good reason to drop them.

I am not proposing that we drop the hashes, they just need to be a separate interface, complimentary to the resumer interface. I quite like the idea of the rolling hashsum, the implementation should just be composable rather than ignorable parameters.

ivandeex replied:
> Do you mean the cache is not stored on the remote?

It's stored locally under ~/.cache/rclone/uploads or whatever

> You can't assume that there is a local filesystem available to rclone.

Indeed rclone keeps configs and vfs cache locally.

> rclone should make effort to support resumable uploads regardless of if the backend supports it or not. This can be done with a configurable fallback backend for intermediate files

Supporting "fallback upload backends" is out of scope for this PR. This can be discussed elsewhere.

ivandeex replied:
> the implementation should just be composable rather than ignorable parameters.

I'd prefer the latter. This is discussible.

innovate-invent commented on fs/operations/resume.go:
```diff
@@ -0,0 +1,178 @@
+
+// readResumeCache checks to see if a resume ID has been cached for the source object.
+// If it finds one it returns it along with true to signal a resume can be attempted
+func readResumeCache(ctx context.Context, f fs.Fs, src fs.Object, cacheName string) (resumeID, hashName, hashState string, attemptResume bool) {
```
This implementation was written with the chunker in mind, but not all backends need to manage uploads as a set of files.

Assuming that there is a cache folder available forces backend resumable uploads into a specific method of managing the uploads.

For example, if I wanted to implement resumable uploads using the multipart upload functionality of the AWS S3 API, then there are no objects created in the backend until the upload is complete.

ivandeex replied:
See above

ivandeex replied:
The implementation is generic enough to handle every suppirted backend.

innovate-invent replied:
The point isn't that some backends are not capable of managing a cache, it is that this forces backends to implement a cache when they might already have better support for resumable uploads via a different mechanism.

ivandeex replied:
> this forces backends to implement a cache

No. They will just return empty hash.

ivandeex replied:
> resumable uploads via a different mechanism

If eg. S3 needs other data to resume an upload, that other data will be added to the signature. If the signature grows above 3-4 parameters, we can factor it and pass a structure.

ivandeex replied:
> the implementation should just be composable rather than ignorable parameters.

I'd prefer the latter. This is discussible.

innovate-invent reviewed (commented):
I am a bit worried this implementation is very biased towards the chunker functionality and places restrictions on how a backend can support resumable uploads.

ivandeex commented:
The interface is generic enough to include features of known backends including S3 or chunker.
Please elaborate what restrictions you think are placed.

innovate-invent commented:
> The interface is generic enough to include features of known backends including S3 or chunker.
> Please elaborate what restrictions you think are placed.

My reservation is that this is more of a framework than an interface. Decisions are being made on how the backend has to implement resumable uploads. In my implementation I was very careful to leave the interface abstract and keep any specific implementation in a separate location that a backend can then import. For example, my [fragmentUploader](https://github.com/innovate-invent/rclone/blob/resumable/fs/resumable/fragmentuploader.go) (which is very similar to your cache) is only one possible implementation available to backends.

ivandeex commented:
> In my implementation I was very careful to leave the interface abstract and keep any specific implementation in a separate location that a backend can then import.

I just care that supported are chunker features that provide for upload safety and leverage chunker-specific on-backend store. It's up to resumer code (which will describe interface, implementation, framework and so on) to make it coherent with S3 or whatever features or choose appropriate level of abstraction. Above I called for @ncw to decide on that.

ivandeex commented:
> my [fragmentUploader](https://github.com/innovate-invent/rclone/blob/resumable/fs/resumable/fragmentuploader.go) (which is very similar to your cache) is `only one possible` implementation available to backends.

:smiley:

innovate-invent commented:
> > my [fragmentUploader](https://github.com/innovate-invent/rclone/blob/resumable/fs/resumable/fragmentuploader.go) (which is very similar to your cache) is `only one possible` implementation available to backends.
> 
> smiley

only one possible, with multiple provided. The backend is free to implement the Uploader interface any way it wants and is not required to use any of the provided implementations.

mcalman commented:
@ivandeex As you suggested I am working on an implementation of Resumer for the local backend so that tests can be written for the interface. Since the local backend supports using multiple hash types at once (with the `MultiHasher`), what should we store in `hashName` and `hashState`?

ivandeex commented:
@mcalman 
Cool! Unit test is the last major sumbling block, I guess.

I suggest using first found hash type (for localfs it will be `md5`) like [object fingerprinter ](https://github.com/rclone/rclone/blob/v1.55.1/fs/fingerprint.go#L45):

```
hashType := f.Hashes().GetOne()
```

For saving `hashState` (intermediate hashing state of an incomplete file) you'll want to extend the [rclone MultiHasher](https://github.com/rclone/rclone/blob/v1.55.1/fs/hash/hash.go#L198) with a function following the pattern in the golang [hash marshaling example ](https://golang.org/pkg/hash/#example__binaryMarshaler) to return a snapshot of partial hashing operation.  Here is a sketch (names are tentative, you can use `hex` instead of `base64` if you like):

```
import "encoding/base64"

func (m *MultiHasher) GetHashState(hashType Type) (string, error) {
	h, ok := m.h[hashType]
	if !ok {
		return "", ErrUnsupported
	}
	snapshotter, ok := h.(encoding.BinaryMarshaler)
	if !ok {
		return "", errors.New("snapshot not implemented")
	}
	data, err := snapshotter.MarshalBinary()
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(data), nil
}
```

You will also need a reverse function that resets MultiHasher (or one of its hashes) to a previous state,
so upload and hashing can resume.

ivandeex commented:
@mcalman
Rebased on new master. Please don't force-push now.

The pre-state of commits has been saved under [ivandeex/mcalman-add-resume-interface](https://github.com/ivandeex/rclone/tree/mcalman-add-resume-interface).

Also rebased #5185 on top of this PR.

mcalman commented:
@ivandeex @ncw As requested I implemented the Resumer interface for the local backend to allow for tests to be written. I also added an initial test that that simulates an upload being interrupted, restarted, and then checks that it was successfully resumed. Please let me know what other tests you think should be added or if you see any problems with the implementation.

ivandeex commented:
@mcalman 

> Please let me know what other tests you think should be added or if you see any problems with the implementation.

I don't have any more requirements.

New tests fail on Windows because you use Unix-only kill signal etc. The simplest and fastest way to mitigate this is to put them in a separate xxx_test.go file with build comment `// +build !windows`. The tests will not be run on Windows. Another way is to use core go APIs from `os/exec` instead of unix-only.

Thank you very much for patience.

@ncw It's your turn to review this patch.

ivandeex commented on lib/atexit/atexit.go:
```diff
@@ -46,12 +46,12 @@ func Register(fn func()) FnHandle {
 				return
 			}
-			signal.Stop(exitChan)
+			//signal.Stop(exitChan)
```
This change will break a lot in rclone.

ivandeex reviewed (commented):
See inline comments

ivandeex commented:
Rebased on latest master to fix merge conflict.

ivandeex commented on fs/operations/resume.go:
```diff
@@ -0,0 +1,200 @@
+	ci := fs.GetConfig(ctx)
+	cacheCleaned := false
+	return func(ID, hashName, hashState string) (err error) {
+		rootCacheDir := filepath.Join(config.CacheDir, "resume")
```
@mcalman 
Sorry. I thought I've finished reviewing but just now I noticed that this place violates a Go convention.
The context pointer "ctx" must always be passed only with call stack, but here it's passed from the resume option setup code in `createSetID` via closure to the returned anonymus function that will be called later by backend, probably in a separate call stack.
I suggest to clarify this place as follows: get rid of the anonymous function with its work data passed via closure. Instead, declare it as a public method of the resume option structure:
```
func (opt *fs.OptionResume) SetID(ctx context.Context, ID, hashName,  hashState string) error {
 ...
}
```
Then call it in backend as `err := resumeOpt.SetID(ctx, id, ...)`.
Pass its state, e.g `cacheCleaned` as a private member of the resume option structure.
Don't pass context or anything derived from it (`ci`) via structure, it's a violation altogether.
Derive it righr from new context in `SetID`.

ivandeex reviewed (commented):
Rebased on latest master to fix merge conflict.
Please pull rebased tree from github.

Also fixed unit test because after commit 6b7d7d0 when rclone is signalled it will now exit with `signum` or `128+signum` code instead of `1` (or `0`).

ivandeex commented on fs/operations/resume_test.go:
```diff
@@ -0,0 +1,165 @@
+				wg.Done()
+				return
+			}
+		}
```
`Signalled` returns immediately so this loop is in fact a short-living 100% cpu hog, a spin-poll. I suggest to add a small delay like `time.Sleep(10*time.Millisecond)` after the `if`.

ivandeex replied:
@mcalman
`Signalled` returns immediately so this loop is in fact a short-living spin-poll, 100% cpu hog.
Generally this means that you'd better revise your solution idea. I am leaving that decision to @ncw and yourself.
As a _quick fix_ I would add a small delay: (?)
```
go func() {
		for {
			if atexit.Signalled() {
				wg.Done()
				return
			}
			time.Sleep(10 * time.Millisecond)
		}
} 
```

ivandeex reviewed (commented):
:(

ivandeex reviewed (approved):
approved

ivandeex reviewed (commented):
approved

ivandeex commented:
This comment is not a requirement but a side note.
I  skimmed through the subthread of #87spanning from https://github.com/rclone/rclone/issues/87#issuecomment-258645510 till https://github.com/rclone/rclone/issues/87#issuecomment-280484648 and compared with API developed by this PR. I have an impression that the API has got the key bits allowing to base yet another backend, the Amazon ACD resume feature upon it:

- ability to set the minimum size requirement for resume
- it's the backend who assigns breakpoint identifiers depending on upstream requirements
- resume id, being an opaque string, can be adapted to the ACD's LocalId/uuid case
- support for intermediate checksums

This does not guarantee that the API and the resume cache machinery will not need fine grinding. On the contrary, adding every other backend to the suit will result in more API refactorings. At least, now I can hope we are on the right way. Nick Craig-Wood for sure will add more comments and  requirements.

The key feature of this PR is that it brings in a passing unit test with it, the first working prototype and consequently a way to play with the feature in practice and watch the "loop closed" from command line to a backend. This will give developers who improve or refactor the interface (add new backends, clarify the logic or implement their favorite programming patterns) a firm ground and a way to check their ideas for sanity.

ivandeex commented:
This comment is not a requirement but a side note.

I  skimmed through the subthread of #87 spanning from https://github.com/rclone/rclone/issues/87#issuecomment-258645510 till https://github.com/rclone/rclone/issues/87#issuecomment-280484648 and compared with API developed by this PR. I have an impression that the API has got the key bits allowing to base yet another backend, the Amazon ACD resume feature upon it (not in this PR, of course, but in future):

- ability to set the minimum size requirement for resume
- it's the backend who assigns breakpoint identifiers depending on upstream requirements, by means of callback
- resume id, being an opaque string, can be adapted to the ACD's LocalId/uuid case
- support for intermediate checksums

This does not guarantee that the API and the resume cache machinery will not need fine grinding. On the contrary, adding every other backend to the suit will result in more API refactorings. I don't even try to predict what will be introduced by 30+ supported rclone backends. Maybe I'd do if I had a budget and 10+ engineers working on this full-time. Even in that case the YAGNI principle ("don't try to do too much ahead of time") might have stopped me. At least, now I can hope that we are on the right way. Nick Craig-Wood for sure will add more comments.

The key feature of this PR is that it brings in a passing unit test with it, the first working prototype and consequently a way to play with the feature in practice and watch the "loop closed" from command line to a backend. This will give developers who improve or refactor the interface (adding new backends, clarifying the logic or implementing their favorite programming patterns) the firm ground and a way to check their ideas for sanity.

ivandeex commented:
@mcalman 

1. I squash-rebased this PR on top of master. Please force-pull the tree from github. Now what we have here is just rclone v1.56.0 + single commit.
2. It's time to start beta-testing of the resulting localfs-based prototype and see whether CLI behavior feels consistent. So I published beta build https://github.com/ivandeex/rclone/releases/tag/v1.56.0-resume01
3. Currently your patch in #5185 lives on top of older version of this PR. I guess you might be interested in rebasing to look how they feel together now, e.g. run unit tests with chunker.

mcalman commented:
@ncw Will you have time to review this soon? I’ll have limited time to work on this PR after next week. If there are any major revisions that need to be made it would be great to be able to work on them soon.

ivandeex commented:
just rebased on master to resolve a conflict after recent master commits, please don't force-push.

ivandeex commented:
@ncw :pray: :pray: :pray:

ncw commented on backend/local/local.go:
```diff
@@ -1338,9 +1399,44 @@ func cleanRootPath(s string, noUNC bool, enc encoder.MultiEncoder) string {
+		return cachedPos, nil
+	}
+	// No valid position found, restart from beginning
+	fs.Debugf(f, "Cached hash state did not match hash state on remote")
```
This should probably be an Info message

```
fs.Infof(remote, "Not resuming as cached hash state did not match hash state on remote")
```

ncw commented on fs/config/configflags/configflags.go:
```diff
@@ -131,6 +131,8 @@ func AddFlags(ci *fs.ConfigInfo, flagSet *pflag.FlagSet) {
 	flags.DurationVarP(flagSet, &ci.FsCacheExpireInterval, "fs-cache-expire-interval", "", ci.FsCacheExpireInterval, "interval to check for expired remotes")
 	flags.BoolVarP(flagSet, &ci.DisableHTTP2, "disable-http2", "", ci.DisableHTTP2, "Disable HTTP/2 in the global transport.")
+	flags.FVarP(flagSet, &ci.MaxResumeCacheSize, "max-resume-cache-size", "", "The maximum size of the cache used to store data necessary for resuming uploads. When the storage grows beyond this size, the oldest resume data will be deleted. (default 100k")
+	flags.FVarP(flagSet, &ci.ResumeLarger, "resume-larger", "", "If set, attempt to resume all partial uploads larger than this size. (default off)")
```
`resume-cutoff` would be more in keeping with the other rclone flags?

ncw commented on fs/config/configflags/configflags.go:
```diff
@@ -131,6 +131,8 @@ func AddFlags(ci *fs.ConfigInfo, flagSet *pflag.FlagSet) {
 	flags.DurationVarP(flagSet, &ci.FsCacheExpireDuration, "fs-cache-expire-duration", "", ci.FsCacheExpireDuration, "cache remotes for this long (0 to disable caching)")
 	flags.DurationVarP(flagSet, &ci.FsCacheExpireInterval, "fs-cache-expire-interval", "", ci.FsCacheExpireInterval, "interval to check for expired remotes")
 	flags.BoolVarP(flagSet, &ci.DisableHTTP2, "disable-http2", "", ci.DisableHTTP2, "Disable HTTP/2 in the global transport.")
+	flags.FVarP(flagSet, &ci.MaxResumeCacheSize, "max-resume-cache-size", "", "The maximum size of the cache used to store data necessary for resuming uploads. When the storage grows beyond this size, the oldest resume data will be deleted. (default 100k")
```
This is `maximum number of items` rather than `maximum size` isn't it?

ivandeex replied:
I recall I was asking to limit by size, not by items. AFAIC recall the author satisfied the request.

ncw commented on fs/hash/hash.go:
```diff
@@ -248,6 +250,67 @@ func (m *MultiHasher) Size() int64 {
+	if !ok {
+		return "", ErrUnsupported
+	}
+	marshaler, ok := h.(encoding.BinaryMarshaler)
```
How many of our hashes implement `encoding.BinaryMarshaler`? I don't think any I've written do (dropboxhash, onedrive hash).

ivandeex replied:
stdlib hashes do like md5, sha1, sha256

ivandeex replied:
I will submit a pr for mrhash

ivandeex replied:
... and for dropbox/onedrive hashes after this pr gets merged

ivandeex replied:
Captured in https://github.com/rclone/rclone/issues/5624

ncw commented on fs/open_options.go:
```diff
@@ -230,6 +236,150 @@ func (o *HashesOption) Mandatory() bool {
+		return errors.Wrapf(err, "failed to marshal data JSON")
+	}
+	if len(data) < int(ci.MaxResumeCacheSize) {
+		// Each remote will have its own directory for cached resume files
```
We should probably factor the code out of the VFS layer which makes a filename for a cache directory based on an Fs. Its needed fixing sooo many times so re-implementing it doesn't seem like a good idea!

ncw commented on fs/operations/interrupt.go:
```diff
@@ -0,0 +1,22 @@
+//+build !windows
```
These should probably go in `interrupt_test.go` as they are only used in the tests aren't they?

ncw commented on fs/operations/operations.go:
```diff
@@ -475,6 +484,12 @@ func Copy(ctx context.Context, f fs.Fs, dst fs.Object, remote string, src fs.Obj
 							newDst = dst
 							err = closeErr
+							// Remove resume cache file (if one was created) when Put/Upload is successful
+							cacheFile := filepath.Join(config.CacheDir, "resume", f.Name(), f.Root(), remote)
```
I think this is the incorrect path with the path mangling done when creating the cache dir?

ncw commented on fs/operations/resume.go:
```diff
@@ -0,0 +1,77 @@
+	if ci.ResumeLarger >= 0 {
+		root := f.Root()
+		if runtime.GOOS == "windows" {
+			if root[:4] == "//?/" {
```
Duplicated cache path mangling!

ncw reviewed (commented):
This looks very nice now :-)

I put a few comments inline for you to look at.

Let's try to get this merged for 1.57 :-)

mcalman commented on lib/file/preallocate_unix.go:
```diff
@@ -47,7 +47,6 @@ func PreAllocate(size int64, out *os.File) (err error) {
 			// Try the next flags combination
 			index++
 			atomic.StoreInt32(&fallocFlagsIndex, index)
-			fs.Debugf(nil, "preAllocate: got error on fallocate, trying combination %d/%d: %v", index, len(fallocFlags), err)
```
I removed this for now because there was an import cycle I couldn't seem to avoid with the `cacheroot`, `file`, and `fs` packages. If there is a way to avoid this or a better place to define the `CreateCacheRoot` function please let me know.

ivandeex replied:
`lib/file` is being modified concurrently by pr https://github.com/rclone/rclone/pull/5401#issuecomment-913845588
and needs attention of @albertony @ncw

ivandeex replied:
probably we could introduce an interface for adhoc open option?

mcalman commented:
@ncw @ivandeex Do we think we can still include this and #5185 in 1.57? Is there anything else that needs to be done?

ivandeex commented:
I just rebased this on the git tip, resolved merge conflicts and remade the cache root tefactoring after recent changes on the master.
It all depends on @ncw review now.

ivandeex commented:
I rebased this on the git tip, resolved merge conflicts, fixed order of CreateCacheRoot arguments in fs/operations.go, remade the cache root refactoring to follow os-safe cache dirs after recent changes on the master.
p
It all depends on @ncw review now.

ivandeex commented:
I rebased this on the git tip, resolved merge conflicts, fixed order of CreateCacheRoot arguments in fs/operations.go, remade the cache root refactoring to follow os-safe cache dirs after recent changes on the master.

please clean up commit messages and author emails.

all depends on @ncw review now.

ivandeex commented:
imho this is ready for 1.57

ivandeex commented:
@mcalman Yet another rebase on new master and resolved conflicts, don't force push..

@ncw PTAL!

ivandeex committed:
local: improve atexit handler for resume

- report errors if any
- prevent double invocation
- prefer saved hash type when many are supported

ivandeex commented:
@ncw
For extra safety, we could use the release strategy similar to bisync:
merge this _right after_ 1.57,
let beta build users test it and submit patches,
and finally have it in 1.58 ofiicially.

ivandeex commented:
@ncw
For extra safety, we could use the release strategy similar to bisync:
- merge this _right after_ 1.57,
- let beta build users test it and submit patches (include mine above and the chunker-resume pr),
- and finally have it in 1.58 ofiicially.

innovate-invent commented:
My objections to this PR seem to have been collapsed under loads of commit messages.
@ncw please take them into consideration before merging
https://github.com/rclone/rclone/pull/4547#issuecomment-831718386

ivandeex commented:
This "loads of commits and messages" is called working on your PR.

ivandeex committed:
Apply go mod tidy to fix build

ivandeex commented:
@mcalman

This PR is waiting for @ncw to review
- #5299

which is a spin off from
- #3850

which is stuck forever. 
Unfortunately that PR is pretty far from production and invasive on core parts of rclone.
Reviewing by reading the code it is not simple and not enough as it like this PR changes code in the rclone core.
The process will definitely include writing tests and ensuring that current functionality does not get broken.

As far as I can see @ncw actively discussed early design of that PR.
Currently waiting for @ncw to find time for review.
If it's not done by next January, I am going to untangle the knot myself.
Probably I will combine code from both proposals, add appropriate tests and submit as new patch.
In the meantime I have added the patch from
- #5185

on top of this one - at any rate I will consider all parts together.

I do not expect the solution to come before the 1.58 release so I expect the resume feature to appear in 1.59.


@innovate-invent This patch will not be merged on rclone master branch before that, which is at least 3 months from now.

However, the patch of this PR has already solved me hours of work time by resuming downloads ("uploads" onto local backend) failed due to network hiccups.
In the meantime I created a branch and a beta release to use this code in practice.
The beta releases will be available at https://beta.rclone.org/branch/resume/

@mcalman By the way, you can open pull requests against the beta branch https://github.com/rclone/rclone/commits/resume

cc @ncw

ivandeex commented:
@mcalman I am sorry about the situation

ivandeex commented:
The PR branch as well as https://beta.rclone.org/branch/resume/ is rebased on the current master.

maxcalman commented:
> If it's not done by next January, I am going to untangle the knot myself.

@ivandeex Was just wondering what your status is. Do you still think it is likely the resume feature will be included in 1.59?

ivandeex commented:
I'm hit heavily by anti-Rus sanctions, plans broken. Most of the short-term time I'll be recovering whatever I can w.r.t. work env, accounts and payments etc. Can't promise anything concrete a.t.m.

ncw commented:
Hi @mcalman, and thank you for this, and for the substantial design discussion that went into it.

I'm doing a tidy-up of long-inactive PRs, and this one has been quiet for around three years. The resumable-upload interface is still an interesting idea, but the code has drifted a long way from current `master` and the design thread never quite converged, so I'm going to close it to keep the queue manageable.

This really isn't a reflection on your effort, as there was a lot of good thinking here. If you (or anyone reading) would like to revive resumable uploads, a fresh PR or a [forum](https://forum.rclone.org/) discussion to settle the design would be very welcome. Thanks again for the work.
